### PR TITLE
System precompiled contracts accounts

### LINKED ISSUE #4126: Create and externalize the precompiled contracts accounts
The following precompiled system contracts exist in Hedera: 
- HTS precompiled contract at `0x167` or account number `0.0.359`
- Exchange rate precompiled contract at `0x168` or account number `0.0.360`
- PRNG system precompiled contract at `0x169` or account number `0.0.361`

Currently, they are only known to the EVM internally and the Hedera nodes have no knowledge of these accounts (even though we have [reserved the ](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/config/HederaNumbers.java#L31)0.0.350-399 id range for precompiled contracts).

Since these precompiles are implemented and exist in Hedera, it makes sense actually to create the accounts. This way:
- the mirror nodes would know about the accounts and return some meaningful information about them
- direct calls to these addresses would be possible and users wouldn't have to deploy contracts that themselves call the precompiles

Some things to consider:
- we should create these contracts with meaningful bytecode 0xFE+SYSTEMCONTRACT (FE→254 invalid instruction for Ethereum)
- these accounts should not be able to receive funds (HBars or tokens)
- these accounts should have a meaningful memo 
- we should handle auto-renewal and expiry for these accounts (they should not expire)

### PR DESCRIPTION
**Pull request by dimitar-dinev**
**NOTE FOR REVIEWERS** 

The 4k file diff is misleading --- 3k of the diff is from [the added `IHederaTokenService.json `in test-clients](https://github.com/hashgraph/hedera-services/pull/5730/files#diff-096bc19ff2261cb43b025cc47ce10500332804e3d59b9bf7882cb919410e62d5R1)

---

**DESCRIPTION**

Adds logic that creates and migrates system precompiled contract accounts, all behind a feature flag.

- Adds 2 new properties:
    - `contracts.createSystemContracts`  — a feature-flag, gating all the logic of system contracts described in the next sections
        - set as `true` in the *src/main/resources* properties, so that CI creates the contracts and the new tests in `DirectPrecompileCallSuite` pass.
        - set as `false` in the specific configurations for for the different environments (mainnet, testnet, preprod, etc)
    - `contracts.systemContracts` — a comma-separated collection of the account nums of the enabled system contracts
        - set as `359,360,361`  (marking `HTS`, `ExchangeRate`, and `PRNG` system contracts as active) in the *src/main/resources* properties, so that CI creates the contracts and the new tests in `DirectPrecompileCallSuite` pass.
        - set as empty in all other the specific configurations for the different environments (mainnet, testnet, preprod, etc)
        - ⚠️ not completely confident this needs to be a property; can be a constant in the code as well;
- Adds a [new SystemContractsCreator class](https://github.com/hashgraph/hedera-services/pull/5730/files#diff-99c039510dad3c91bfab7b7ac07c2b58f46fbbec7608d144408299d880099c0aR55), which encapsulates all of the logic for the creation of the system precompiled contracts. It [creates an account in the state for each system contract num marked](https://github.com/hashgraph/hedera-services/pull/5730/files#diff-99c039510dad3c91bfab7b7ac07c2b58f46fbbec7608d144408299d880099c0aR83-R106) in the `contracts.systemContracts` property.
    - Each system contract is created with:
        - ******key****** — STANDIN-CONTRACT-ID-KEY
        - **expiry** — the system entity expiry returned by the `bootstrap.system.entityExpiry` properrty
        - **autoRenewPeriod** — the system entity expiry returned by the `bootstrap.system.entityExpiry` property
        - **receiverSigRequired** — true
        - **declineReward** — true
        - **bytecode** - as per https://github.com/hashgraph/hedera-services/issues/4126#issuecomment-1429890670
- The `SystemContractsCreator` is plugged into 3 places, where **all logic is gated by the `contracts.createSystemContracts`**:
    - In [BackedSystemAccountsCreator](https://github.com/hashgraph/hedera-services/pull/5730/files#diff-0aa7c5f48ea2bcc02ce15efae98fb5d350966487e01ab7efc1fe3056ae513923R127-R130)  — make sure calls to `ensureSystemAccounts()` during `GENESIS` init also create  the system contracts. Also, acts as a proxy of getting the created system contracts.
    - in [ServicesState#migrateFrom](https://github.com/hashgraph/hedera-services/pull/5730/files#diff-bf6562511716122b40a2e384789c7e81750fd389c914be741af41a2997a40abdR613-R616)  — apart from creating the system contracts on `GENESIS`, make sure they are also created in the next major release
    - in [MigrationRecordsManager](https://github.com/hashgraph/hedera-services/pull/5730/files#diff-ed7cc4b674e078cd9fde7e887fb965c038c353b52ac525e82573e9dff9370480R147-R150) — export synthetic `ContractCreate` transactions for all created system contracts
- Adds a new `DirectPrecompileCallsSuite`
    - More tests to be added in subsequent PRs 
---

**Related issue(s)**:

Fixes #4126

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                    continue;
+                }
+                final var htsPrecompiledContract = new HederaAccountCustomizer()
+                        .isSmartContract(true)
+                        .isReceiverSigRequired(true)
+                        .isDeclinedReward(true)
+                        .isDeleted(false)
+                        .expiry(expiry)
+                        .memo(SYSTEM_CONTRACT_MEMO)
+                        .key(STANDIN_CONTRACT_ID_KEY)
```
Comment (dimitar-dinev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java:101]: Or maybe genesis key 🤔

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        final var expiry = properties.getLongProperty(BOOTSTRAP_SYSTEM_ENTITY_EXPIRY);
+        final var activeSystemContracts = properties.getAccountNums(CONTRACTS_SYSTEM_CONTRACTS);
+        for (long num = HederaNumbers.FIRST_RESERVED_SYSTEM_CONTRACT;
+                num <= HederaNumbers.LAST_RESERVED_SYSTEM_CONTRACT;
+                num++) {
+            final var id = STATIC_PROPERTIES.scopedAccountWith(num);
+            if (activeSystemContracts.contains(id.getAccountNum())) {
+                if (accounts.contains(id)) {
+                    continue;
+                }
```
Comment (dimitar-dinev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java:93]: Can directly iterate through the `activeSystemContracts` Set

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ServicesApp.java [hunk 3]
```diff
@@ -208,6 +209,8 @@ public interface ServicesApp {
 
     BackingStore<AccountID, HederaAccount> backingAccounts();
 
+    SystemContractsCreator systemContractsCreator();
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ServicesApp.java:212]: Javadoc is missing

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/context/properties/PropertySource.java [hunk 4]
```diff
@@ -195,6 +198,10 @@ default Set<Address> getEvmAddresses(String name) {
         return getTypedProperty(Set.class, name);
     }
 
+    default Set<Long> getAccountNums(String name) {
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/context/properties/PropertySource.java:201]: Javadoc is missing.
null checks and annotations are missing

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/BackedSystemAccountsCreator.java [hunk 5]
```diff
@@ -68,12 +70,14 @@ public BackedSystemAccountsCreator(
             final @CompositeProps PropertySource properties,
             final Supplier<JEd25519Key> genesisKeySource,
             final Supplier<HederaAccount> accountSupplier,
-            final TreasuryCloner treasuryCloner) {
+            final TreasuryCloner treasuryCloner,
+            final SystemContractsCreator systemContractsCreator) {
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/BackedSystemAccountsCreator.java:74]: null annotations and null checks missing

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/BackedSystemAccountsCreator.java [hunk 6]
```diff
@@ -171,12 +178,17 @@ public List<HederaAccount> getSystemAccountsCreated() {
         return systemAccountsCreated;
     }
 
+    public List<HederaAccount> getSystemContractsCreated() {
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/BackedSystemAccountsCreator.java:181]: null annotations missing

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * <p>As of release 0.37.0, there are 3 such contracts:
+ * <ul>
+ *     <li>{@link HTSPrecompiledContract}</li>
+ *     <li>{@link ExchangeRatePrecompiledContract}</li>
+ *     <li>{@link PrngSystemPrecompiledContract}</li>
+ * </ul>
+ */
+@Singleton
+public class SystemContractsCreator {
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java:55]: This creates a lot of new functionality in the mono-service module that we want to remove in the long run. Can we move the new functionality to another module?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    private final List<HederaAccount> contractsCreated = new ArrayList<>();
+    private final PropertySource properties;
+    private final EntityAccess entityAccess;
+
+    @Inject
+    public SystemContractsCreator(
+            final Supplier<HederaAccount> accountSupplier,
+            final BackingStore<AccountID, HederaAccount> accounts,
+            final @CompositeProps PropertySource properties,
+            final EntityAccess entityAccess) {
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java:72]: null annotations and null checks missing

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        }
+    }
+
+    /**
+     * Returns a list of {@link HederaAccount}, denoting all of the system contracts created
+     * by a previous call to {@link SystemContractsCreator#ensureSystemContractsExist()}
+     *
+     * @return a list of all the created system contracts
+     */
+    public List<HederaAccount> getContractsCreated() {
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java:117]: null annotations  missing

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/MigrationRecordsManager.java [hunk 10]
```diff
@@ -220,6 +229,56 @@ private TransactionBody.Builder synthCreation(
...
+        tracker.trackAutoCreation(num.toGrpcAccountId());
+        final var synthBody = synthContractCreation(autoRenewPeriod, key, accountMemo, receiverSigRequired, balance);
+        final var synthRecord =
+                creator.createSuccessfulSyntheticRecord(NO_CUSTOM_FEES, tracker, SYSTEM_CONTRACT_CREATION_MEMO);
+
+        recordsHistorian.trackPrecedingChildRecord(DEFAULT_SOURCE_ID, synthBody, synthRecord);
+        sigImpactHistorian.markEntityChanged(num.longValue());
+        log.info(
+                "Published synthetic ContractCreate for {} account 0.0.{}",
+                "system contract creation",
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/migration/MigrationRecordsManager.java:260]: why is a string a param?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java (hunk 1 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/initialization/SystemContractsCreator.java:101]: If we are treating them as system accounts, we should use genesis key.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ServicesState.java [hunk 11]
```diff
@@ -610,6 +610,10 @@ void migrateFrom(@NonNull final SoftwareVersion deserializedVersion) {
         // Keep the MutableStateChildren up-to-date (no harm done if they are already are)
         final var app = getMetadata().app();
         app.workingState().updatePrimitiveChildrenFrom(this);
+        if (app.bootstrapProps().getBooleanProperty(PropertyNames.CONTRACTS_CREATE_SYSTEM_CONTRACTS)
+                && SEMANTIC_VERSIONS.deployedSoftwareVersion().hasMigrationRecordsFrom(deserializedVersion)) {
+            app.systemContractsCreator().ensureSystemContractsExist();
+        }
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ServicesState.java:616]: I think this should be at similar place as `ensureSystemAccounts`

**Comment by dimitar-dinev** [comment]
Closing since this approach has been replaced by a new one described in #5986.

**Pull request closed without merging**